clang/lib/CIR/Dialect/Transforms/CIRSimplify.cpp

```diff
@@ -107,6 +111,91 @@ struct RemoveTrivialTry : public OpRewritePattern<TryOp> {
+/// - The true branch and the false branch cannot be "too costly" since both of
+///   them will be executed after the folding happens.
+///
+/// For now we only simplify those ternary operations whose true and false
+/// branches directly yield a value or a constant. That is, both of the true and
+/// the false branch must either contain a cir.yield operation as the only
+/// operation in the branch, or contain a cir.const operation followed by a
+/// cir.yield operation that yields the constant value.
+///
+/// For example, we will simplify the following ternary operation:
+///
+///   %0 = cir.ternary (%condition, true {
+///     %1 = cir.const ...
+///     cir.yield %1
+///   } false {
+///     cir.yield %2
+///   })
+///
+/// into the following sequence of operations:
+///
+///   %1 = cir.const ...
+///   %0 = cir.select if %condition then %1 else %2
+struct SimplifyTernary final : public OpRewritePattern<TernaryOp> {
+  using OpRewritePattern<TernaryOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TernaryOp op,
+                                PatternRewriter &rewriter) const override {
+    if (op->getNumResults() != 1)
+      return mlir::failure();
+
+    if (!isTernaryBranchSimplifiable(op.getTrueRegion()) ||
+        !isTernaryBranchSimplifiable(op.getFalseRegion()))
+      return mlir::failure();
+
+    mlir::cir::YieldOp trueBranchYieldOp = mlir::cast<mlir::cir::YieldOp>(
+        op.getTrueRegion().front().getTerminator());
+    mlir::cir::YieldOp falseBranchYieldOp = mlir::cast<mlir::cir::YieldOp>(
+        op.getFalseRegion().front().getTerminator());
+    auto trueValue = trueBranchYieldOp.getArgs()[0];
+    auto falseValue = falseBranchYieldOp.getArgs()[0];
+
+    rewriter.inlineBlockBefore(&op.getTrueRegion().front(), op);
+    rewriter.inlineBlockBefore(&op.getFalseRegion().front(), op);
+    trueBranchYieldOp.erase();
+    falseBranchYieldOp.erase();
+    rewriter.replaceOpWithNewOp<mlir::cir::SelectOp>(op, op.getCond(),
+                                                     trueValue, falseValue);
+
+    return mlir::success();
+  }
+
+private:
+  bool isTernaryBranchSimplifiable(mlir::Region &region) const {
+    if (!region.hasOneBlock())
+      return false;
+
+    mlir::Block &onlyBlock = region.front();
+    auto &ops = onlyBlock.getOperations();
+
+    // The region/block could only contain at most 2 operations.
+    if (ops.size() > 2)
+      return false;
+
+    if (ops.size() == 1) {
+      // The region/block only contain a cir.yield operation.
+      return true;
+    }
+
+    // Check whether the region/block contains a cir.const followed by a
+    // cir.yield that yields the value.
+    auto yieldOp = mlir::cast<mlir::cir::YieldOp>(onlyBlock.getTerminator());
+    auto yieldValueDefOp = mlir::dyn_cast_if_present<mlir::cir::ConstantOp>(
+        yieldOp.getArgs()[0].getDefiningOp());
+    if (!yieldValueDefOp || yieldValueDefOp->getBlock() != &onlyBlock)
```

Comment:
I'm not sure I get what's the point with the block comparison, if the constantop is defined outside, why would that even be a problem or make the transformation invalid? (It can't happen right now because in such cases, which would still be valid, we early exit with true in `ops.size() == 1` above).
Please remove this block logic.

Reply:
Is the following CIR possible now?

```mlir
cir.ternary (%0, true {
  cir.call ...  // Or any other operation other than cir.const
  cir.yield %1
}, false {
  // ...
})
```

I was concerning about this case when implementing this.

Reply:
Gotcha, checking for the constanop is legit! We don't seem to need the block stuff though?

Reply:
> We don't seem to need the block stuff though?

How could we detect the case above (a.k.a. the branch contains two operations but the first operation is not a `cir.const`) then?

Reply:
Oh right, true!